planner/core/optimizer.go

```diff
@@ -720,58 +883,84 @@ func setupFineGrainedShuffleInternal(plan PhysicalPlan, helper *fineGrainedShuff
 			// Global sort will break the data partition.
 			helper.clear()
 		}
-		setupFineGrainedShuffleInternal(x.children[0], helper, streamCount)
+		setupFineGrainedShuffleInternal(ctx, sctx, x.children[0], helper, streamCountInfo, tiflashServerCountInfo)
 	case *PhysicalSelection:
 		helper.plans = append(helper.plans, &x.basePhysicalPlan)
-		setupFineGrainedShuffleInternal(x.children[0], helper, streamCount)
+		setupFineGrainedShuffleInternal(ctx, sctx, x.children[0], helper, streamCountInfo, tiflashServerCountInfo)
 	case *PhysicalProjection:
 		helper.plans = append(helper.plans, &x.basePhysicalPlan)
-		setupFineGrainedShuffleInternal(x.children[0], helper, streamCount)
+		setupFineGrainedShuffleInternal(ctx, sctx, x.children[0], helper, streamCountInfo, tiflashServerCountInfo)
 	case *PhysicalExchangeReceiver:
 		helper.plans = append(helper.plans, &x.basePhysicalPlan)
-		setupFineGrainedShuffleInternal(x.children[0], helper, streamCount)
+		setupFineGrainedShuffleInternal(ctx, sctx, x.children[0], helper, streamCountInfo, tiflashServerCountInfo)
 	case *PhysicalHashAgg:
-		// HashAgg is not implemented for now.
-		helper.clear()
-		setupFineGrainedShuffleInternal(x.children[0], helper, streamCount)
+		aggHelper := fineGrainedShuffleHelper{shuffleTarget: hashAgg, plans: []*basePhysicalPlan{}}
```

Comment:
agg don't need to use a new `aggHelper`, just reuse `helper` is enough. I think we can leave a todo here.

Reply:
Done